src/execution_plan/optimizations/optimizer.c

```diff
@@ -48,3 +48,13 @@ void optimizePlan(ExecutionPlan *plan) {
 	reduceCount(plan);
 }
 
+void optimizePlan(ExecutionPlan *plan) {
+	AST *ast = QueryCtx_GetAST();
+
+	/* Handle UNION if there are any. */
+	if(AST_ContainsClause(ast, CYPHER_AST_UNION)) {
+		for(uint i = 0; i < plan->segment_count; i++) _optimizePlan(plan->segments[i]);
+	} else {
+		_optimizePlan(plan);
+	}
```

Comment:
Does `plan->segments[0]` points to `plan` by any chance?
if so all we need is: `for(uint i = 0; i < plan->segment_count; i++) _optimizePlan(plan->segments[i]);`

Reply:
Also are there other situations where `plan->segment_count > 0` except UNION?
How would you define an execution plan segment?

Reply:
in `UNION` situations, `segments` do not hold the plan itself
in regular situations, the plan is in index `segments_count +1`


from `NewExecutionPlan`
```
for(int i = 0; i < segment_count; i++) {
		uint end_offset = segment_indices[i];
		// Slice the AST to only include the clauses in the current segment.
		AST *ast_segment = AST_NewSegment(ast, start_offset, end_offset);
		ast_segments[i] = ast_segment;
		// Construct a new ExecutionPlanSegment.
		ExecutionPlan *segment = ExecutionPlan_NewEmptyExecutionPlan();
		ExecutionPlan_PopulateExecutionPlan(segment);

		segments[i] = segment;
		start_offset = end_offset;
	}
```
and
```
	// Disregard self.
	plan->segment_count = segment_count - 1;
	plan->segments = segments;
```